sdkit/filter/codeformer.py

```diff
@@ -0,0 +1,240 @@
+from basicsr.utils.download_util import load_file_from_url
+from sdkit.modules.codeformer.utils.misc import gpu_is_available, get_device
+from sdkit.modules.codeformer.utils.realesrgan_utils import RealESRGANer
+from sdkit.modules.codeformer.utils.registry import ARCH_REGISTRY
+from sdkit.modules.codeformer.utils.logger import get_root_logger
+
+from sdkit.modules.codeformer.archs.codeformer_arch import CodeFormer
+from sdkit.modules.codeformer.facelib.utils.face_restoration_helper import FaceRestoreHelper
+from sdkit.modules.codeformer.facelib.utils.misc import is_gray
+
+import traceback
+import warnings
+import sys
+
+"""
+def warn_with_traceback(message, category, filename, lineno, file=None, line=None):
+
+    log = file if hasattr(file,'write') else sys.stderr
+    traceback.print_stack(file=log)
+    log.write(warnings.formatwarning(message, category, filename, lineno, line))
+
+warnings.showwarning = warn_with_traceback
+"""
+
+def CodeFormer_Initialize():
+    pretrain_model_url = {
+        'codeformer': 'https://github.com/sczhou/CodeFormer/releases/download/v0.1.0/codeformer.pth',
+        'detection': 'https://github.com/sczhou/CodeFormer/releases/download/v0.1.0/detection_Resnet50_Final.pth',
+        'parsing': 'https://github.com/sczhou/CodeFormer/releases/download/v0.1.0/parsing_parsenet.pth',
+        'realesrgan': 'https://github.com/sczhou/CodeFormer/releases/download/v0.1.0/RealESRGAN_x2plus.pth'
+    }
+
+    # download weights
+    for model_type in pretrain_model_url:
+        if not os.path.exists(f'CodeFormer/weights/{model_type}/{model_type}.pth'):
+            load_file_from_url(url=pretrain_model_url[model_type], model_dir=f'CodeFormer/weights/{model_type}', progress=True, file_name=None)
+
+    def imread(img):
+        img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
+        return img
+
+    def set_realesrgan():
+        half = True if gpu_is_available() else False
+        model = RRDBNet(
+            num_in_ch=3,
+            num_out_ch=3,
+            num_feat=64,
+            num_block=23,
+            num_grow_ch=32,
+            scale=2,
+        )
+        upsampler = RealESRGANer(
+            scale=2,
+            model_path="CodeFormer/weights/realesrgan/RealESRGAN_x2plus.pth",
+            model=model,
+            tile=400,
+            tile_pad=40,
+            pre_pad=0,
+            half=half,
+        )
+        return upsampler
+
+    upsampler = set_realesrgan()
+    device = get_device()
+    print('device:', device)
+    codeformer_net = ARCH_REGISTRY.get("CodeFormer")(
+        dim_embd=512,
+        codebook_size=1024,
+        n_head=8,
+        n_layers=9,
+        connect_list=["32", "64", "128", "256"],
+    ).to(device)
+    ckpt_path = "CodeFormer/weights/CodeFormer/codeformer.pth"
+    checkpoint = torch.load(ckpt_path)["params_ema"]
+    codeformer_net.load_state_dict(checkpoint)
+    codeformer_net.eval()
+
+    return imread, set_realesrgan, upsampler, device, codeformer_net, upsampler
```

Comment:
This `CodeFormer_Initialize()` function is loading the model into memory for each image processed. The other filters have this kind of initialization code in the model_loader functions, not in the `apply()` method.

Reply:
Fixed, thanks!

Reply:
Well, not quite. Let me have a closer look tonight.

Reply:
Okay, I think I got it right this time.